qt/python/mantidqt/project/recovery/projectrecovery.py

```diff
@@ -0,0 +1,471 @@
+
+        # Actually execute the current tab
+        QMetaObject.invokeMethod(self.multi_file_interpreter, "execute_current_async_blocking",
+                                 Qt.AutoConnection)
+
+        # Force program to process events so the invoked method is called
+        QApplication.processEvents()
+
+    ######################################################
+    #  Checkpoint Repair
+    ######################################################
+
+    def remove_oldest_checkpoints(self):
+        paths = self.listdir_fullpath(self.recovery_directory_pid)
+
+        if len(paths) > self.maximum_num_checkpoints:
+            # Order paths in reverse and remove the last folder
+            paths.sort(reverse=True)
+            for ii in range(self.maximum_num_checkpoints, len(paths)):
+                self._remove_directory_and_directory_trees(paths[ii])
+
+    def clear_all_unused_checkpoints(self, pid_dir=None):
+        if pid_dir is None:
+            # If no pid directory given then remove user level folder if multiple users else remove workbench-recovery
+            # folder
+            if len(os.listdir(self.recovery_directory)) == 1 and len(os.listdir(self.recovery_directory_hostname)) == 0:
+                path = self.recovery_directory
+            else:
+                path = self.recovery_directory_hostname
+        else:
+            path = pid_dir
+        if os.path.exists(path):
+            try:
+                self._remove_directory_and_directory_trees(path)
+            except Exception as e:
+                if isinstance(e, KeyboardInterrupt):
+                    raise
+                else:
+                    # Fail silently because it is just over diligent calls to clear checkpoints
+                    pass
+
+    def repair_checkpoints(self):
+        """
+        Will remove all locked, older than a month, and empty checkpoints
+        """
+        pid_dirs = self.listdir_fullpath(self.recovery_directory_hostname)
+
+        dirs_to_delete = []
+
+        dirs_to_delete += self._find_checkpoints_older_than_a_month(pid_dirs)
+        dirs_to_delete += self._find_checkpoints_which_are_locked(pid_dirs)
+
+        for dir in dirs_to_delete:
+            self._remove_directory_and_directory_trees(dir)
+
+        # Now the checkpoints have been deleted we may have PID directories with no checkpoints present so delete them
+        self._remove_empty_folders_from_dir(self.recovery_directory_hostname)
+
+    def _find_checkpoints_older_than_a_month(self, pid_dirs):
+        old_pids = []
+        for pid_dir in pid_dirs:
+            # If pid folder hasn't been touched in a month delete it
+            if self._is_file_older_than_month(pid_dir):
+                old_pids.append(pid_dir)
+        return old_pids
+
+    @staticmethod
+    def _is_file_older_than_month(filepath):
+        last_modified = os.path.getmtime(filepath)
+        return last_modified < time.time() - 2592000  # (30 days * 86400 seconds)
```

Comment:
Can we define this "magic" number as a module-level variable called something like `THIRTY_DAYS_IN_SECONDS`? It would aid readability and we wouldn't need the comment.